### PR TITLE
Fix update endpoints

### PR DESCRIPTION
**Pull request by welpaolo**
# Issue
Ready-only endpoints were not updated if a unit joined or left the cluster.

# Solution
Check peer relations joined and departed on the relation to manage the update of the endpoints. A special case is the change of leadership that is handled on leader_elected hook.

# Context
When adding or removing a unit of MySQL the endpoints were not updated.

# Testing
Integration tests includes tests to check the correct update of the endpoints.

### CODE REVIEW

File: src/relations/database.py [hunk 1]
```diff
@@ -43,6 +43,135 @@ def __init__(self, charm):
...
+
+        # check if the leader is departing
+        if self.charm.unit.name == event.departing_unit.name:
+            event.defer()
+            return
+
+        # get unit name that departed
+        dep_unit_name = event.departing_unit.name.replace("/", "-")
+
+        # differ if the added unit is still in the cluster
```
Comment (shayancanonical) [inline src/relations/database.py:95]: differ -> defer in comment

File: src/relations/database.py [hunk 2]
```diff
@@ -43,6 +43,135 @@ def __init__(self, charm):
...
+        if len(relations) == 0:
+            return
+
+        if not self.charm.cluster_initialized:
+            logger.debug("Waiting cluster to be initialized")
+            return
+
+        # check if the leader is departing
+        if self.charm.unit.name == event.departing_unit.name:
+            event.defer()
```
Comment (shayancanonical) [inline src/relations/database.py:89]: do we need to defer here? if the leader is the departing unit, may be we can just return, and let the `leader-elected` handler above handle the update of the endpoints

File: src/relations/database.py [hunk 3]
```diff
@@ -43,6 +43,135 @@ def __init__(self, charm):
...
+            return
+
+        if not self.charm.cluster_initialized:
+            logger.debug("Waiting cluster to be initialized")
+            return
+
+        # get unit name that joined
+        event_unit_label = event.unit.name.replace("/", "-")
+
+        # differ if the added unit is not in the cluster
```
Comment (shayancanonical) [inline src/relations/database.py:128]: differ -> defer in comments

File: tests/integration/helpers.py [hunk 4]
```diff
@@ -372,3 +373,142 @@ def cluster_name(unit: Unit, model_name: str) -> str:
...
+            if "read-only-endpoints" in j_data:
+                read_only_endpoints = j_data["read-only-endpoints"]
+                if read_only_endpoints.strip() == "":
+                    continue
+                for ep in read_only_endpoints.split(","):
+                    roe.add(ep)
+        except json.JSONDecodeError:
+            raise ValueError("Relation data are not valid JSON.")
+
+    return roe
```
Comment (shayancanonical) [inline tests/integration/helpers.py:438]: minor nit -> thoughts on renaming variables so they are named after what they represent? e.g. roe -> read_only_endpoints. ep -> endpoint. r_data -> relation_data, etc

I believe that this makes it easier to read the code for someone who is new to the codebase

File: tests/integration/helpers.py [hunk 5]
```diff
@@ -372,3 +373,142 @@ def cluster_name(unit: Unit, model_name: str) -> str:
...
+    leader_unit = None
+    for app_unit in ops_test.model.applications[application_name].units:
+        is_leader = await app_unit.is_leader_from_status()
+        if is_leader:
+            leader_unit = app_unit.name
+
+    units_to_destroy = [leader_unit]
+
+    for unit_to_destroy in units_to_destroy:
+        await ops_test.model.destroy_units(unit_to_destroy)
```
Comment (shayancanonical) [inline tests/integration/helpers.py:475]: why not just `await ops_test.model.destroy_units(leader_unit)`?

File: tests/integration/test_database.py [hunk 6]
```diff
@@ -241,6 +255,84 @@ async def test_relation_creation(ops_test: OpsTest):
...
+    # check that endpoints are the one of the application
+    for r_endpoint in read_only_endpoints:
+        assert r_endpoint in app_hostnames
+
+    # decrease the number of units
+    async with ops_test.fast_forward():
+        await scale_application(ops_test, DATABASE_APP_NAME, 2)
+
+    # wait for the update of the endpoints
+    time.sleep(2 * 60)
```
Comment (shayancanonical) [inline tests/integration/test_database.py:296]: we generally prefer not to sleep in tests as this is not deterministic. you can instead follow a retry pattern (with tenacity) where you retry if you attempt to get something and it is not present, or use `ops_test.block_until` until a condition is met

[here](https://github.com/canonical/mysql-k8s-operator/blob/2c0c7c5647f3fdd5d7a6e81f5a006ec1f2f9283b/tests/integration/test_charm.py#L110-L125) is an example of the retry pattern

File: src/relations/database.py [hunk 7]
```diff
@@ -43,6 +43,135 @@ def __init__(self, charm):
...
+            return
+
+        relation_data = self.database.fetch_relation_data()
+        # for all relations update the read-only-endpoints
+        for relation in relations:
+            # check if the on_database_requested has been executed
+            if relation.id not in relation_data:
+                logger.debug("On database requested not happened yet! Nothing to do in this case")
+                continue
+            self._update_endpoints(relation.id, self.charm.unit.name)
```
Comment (paulomach) [inline src/relations/database.py:72]:
```suggestion
            self._update_endpoints(relation.id, self.charm.app.name)
```

File: src/relations/database.py (hunk 1 above)
Comment (paulomach) [inline src/relations/database.py:95]: `s/differ/defer`

File: src/relations/database.py [hunk 8]
```diff
@@ -43,6 +43,135 @@ def __init__(self, charm):
...
+                    primary_endpoint,
+                }
+            )
+            self.database.set_read_only_endpoints(relation_id, ",".join(read_only_endpoints))
+            logger.debug(f"Updated endpoints for {remote_app}")
+
+        except MySQLCreateApplicationDatabaseAndScopedUserError:
+            logger.error(f"Failed to create scoped user for app {remote_app}")
+            self.charm.unit.status = BlockedStatus("Failed to create scoped user")
+        except MySQLGetMySQLVersionError as e:
```
Comment (paulomach) [inline src/relations/database.py:166]: Same as above

File: tests/integration/helpers.py [hunk 9]
```diff
@@ -372,3 +373,142 @@ def cluster_name(unit: Unit, model_name: str) -> str:
...
 
     return output[unit.name]["relation-info"][0]["application-data"]["cluster-name"]
+
+
+async def get_relation_data(
+    ops_test: OpsTest,
+    application_name: str,
+    relation_name: str,
+) -> list:
+    """Returns a that contains the relation-data.
```
Comment (paulomach) [inline tests/integration/helpers.py:383]:
```suggestion
    """Returns a list that contains the relation-data.
```

File: tests/integration/helpers.py [hunk 10]
```diff
@@ -372,3 +373,142 @@ def cluster_name(unit: Unit, model_name: str) -> str:
...
+
+    Args:
+        ops_test: The ops test framework instance
+        application_name: The name of the application
+        relation_name: name of the relation to get connection data from
+    Returns:
+        a dictionary that contains the relation-data
```
Comment (paulomach) [inline tests/integration/helpers.py:390]:
```suggestion
        a list that contains the relation-data
```

File: tests/integration/helpers.py [hunk 11]
```diff
@@ -372,3 +373,142 @@ def cluster_name(unit: Unit, model_name: str) -> str:
...
+    """Retrieves hostnames of given application units.
+
+    Args:
+        ops_test: The ops test framework instance
+        app_name: The name of the application
+    Returns:
+        a list that contains the hostnames of a given application
+    """
+    units = [app_unit.name for app_unit in ops_test.model.applications[app_name].units]
+    status = await ops_test.model.get_status()  # noqa: F821
```
Comment (paulomach) [inline tests/integration/helpers.py:501]: why the `noqa` here?

**Review by paulomach — COMMENTED** [review]
Some comments all around

File: tests/integration/helpers.py (hunk 11 above)
Comment (welpaolo) [reply tests/integration/helpers.py:501]: There was a problem with lint on the zookeeper charm. I checked that there is no problem in the charm so I will remove it.

File: src/relations/database.py (hunk 1 above)
Comment (welpaolo) [reply src/relations/database.py]: Done

File: src/relations/database.py (hunk 2 above)
Comment (welpaolo) [reply src/relations/database.py]: Good catch! I changed it

File: src/relations/database.py (hunk 3 above)
Comment (welpaolo) [reply src/relations/database.py]: Done

File: tests/integration/helpers.py (hunk 4 above)
Comment (welpaolo) [reply tests/integration/helpers.py]: I renamed the variables.

File: tests/integration/helpers.py (hunk 5 above)
Comment (welpaolo) [reply tests/integration/helpers.py]: Sure!

File: src/relations/database.py (hunk 7 above)
Comment (welpaolo) [reply src/relations/database.py]: Done

File: src/relations/database.py (hunk 1 above)
Comment (welpaolo) [reply src/relations/database.py]: Done

File: src/relations/database.py [hunk 12]
```diff
@@ -43,6 +43,135 @@ def __init__(self, charm):
...
+            read_only_endpoints = sorted(
+                self.charm._mysql.get_cluster_members_addresses()
+                - {
+                    primary_endpoint,
+                }
+            )
+            self.database.set_read_only_endpoints(relation_id, ",".join(read_only_endpoints))
+            logger.debug(f"Updated endpoints for {remote_app}")
+
+        except MySQLCreateApplicationDatabaseAndScopedUserError:
```
[the remark this answers is not in the document]
Comment (welpaolo) [reply src/relations/database.py]: Exception removed

File: src/relations/database.py (hunk 8 above)
Comment (welpaolo) [reply src/relations/database.py]: Exception removed

File: tests/integration/helpers.py (hunk 10 above)
Comment (welpaolo) [reply tests/integration/helpers.py]: Corrected it

File: tests/integration/helpers.py [hunk 13]
```diff
@@ -372,3 +373,142 @@ def cluster_name(unit: Unit, model_name: str) -> str:
...
+    relation_data = [v for v in data[unit_name]["relation-info"] if v["endpoint"] == relation_name]
+    if len(relation_data) == 0:
+        raise ValueError(
+            f"no relation data could be grabbed on relation with endpoint {relation_name}"
+        )
+
+    return relation_data
+
+
+def get_read_only_endpoint(relation_data: list) -> Set[str]:
```
[the remark this answers is not in the document]
Comment (welpaolo) [reply tests/integration/helpers.py]: Done

**Review by paulomach — APPROVED** [review]
lgtm

**Review by shayancanonical — APPROVED** [review]

**Pull request merged**